src/task.rs

```diff
@@ -480,81 +742,186 @@ enum State {
 }
 
 enum NextState {
     EvaluateTopLevel(String),
+    HandleChildFinished(TaskId),
+    HandleChildErrored(TaskId, Error),
     NotifyStarlingFinished,
     NotifyStarlingErrored(Error),
     NotifyParentFinished,
     NotifyParentErrored(Error),
+    ShutdownChildren,
 }
 
 impl Future for Task {
     type Item = ();
     type Error = Error;
 
     fn poll(&mut self) -> Result<Async<Self::Item>> {
+        // Principals for error handling, recovery, and propagation when driving
+        // tasks to completion:
+        //
+        // * Whenever possible, catch errors and then inform
+        //     1. the Starling system supervisor thread,
+        //     2. the parent task (if any),
+        //     3. and this task's children
+        //   in that order to propagate them.
+        //
+        // * Tasks should never outlive the Starling system supervisor thread,
+        //   so don't attempt to catch errors sending messages to it. Just panic
+        //   if that fails, since something very wrong has happened.
+        //
+        // * We can't rely on this task's children or parent stricly observing
+        //   the tree lifetime discipline that we expose to JavaScript. There
+        //   can be races where both shut down at the same time, perhaps
+        //   involving the main task returning and the Starling system
+        //   supervisor thread sending shutdown notices to all remaining
+        //   tasks. Therefore, we allow all inter-task messages to fail
+        //   silently, locally ignoring errors, and relying on the Starling
+        //   system supervisor thread to perform ultimate clean up.
+
         let next_state = match self.state {
             State::Created => {
                 return self.read_js_module();
             }
             State::ReadingJsModule(ref mut reading) => {
-                let src = try_ready!(reading.poll());
-                NextState::EvaluateTopLevel(src)
+                match reading.poll() {
+                    Err(e) => NextState::NotifyStarlingErrored(e),
+                    Ok(Async::NotReady) => return Ok(Async::NotReady),
+                    Ok(Async::Ready(src)) => NextState::EvaluateTopLevel(src),
+                }
             }
             State::WaitingOnPromise(ref mut promise) => {
-                match try_ready!(promise.poll()) {
-                    Err(_rejection) => {
-                        let err = ErrorKind::JavaScriptUnhandledRejectedPromise;
-                        let err = err.into();
+                // The task is waiting on a promise, but we need to also listen
+                // for messages from our parent, the starling supervisor thread,
+                // or our children. So we need to select from either the promise
+                // we're waiting on or our channel, and depending which wins the
+                // race we need to handle it differently, so we have this little
+                // type to let us know which won the race.
```

Comment:
This means waiting for the promise we're awaiting in the task's main function, right? We could have multiple promises pending, but we can only be awaiting a single one in the main function, right? Would be nice to be explicit about that here.

Reply:
Yes, will clarify.